core/src/mender-api.c

```diff
@@ -90,38 +95,39 @@ mender_api_is_authenticated(void) {
 mender_err_t
 mender_api_perform_authentication(void) {
     mender_err_t       ret;
-    char              *public_key_pem       = NULL;
-    cJSON             *json_identity        = NULL;
-    mender_identity_t *identity             = NULL;
-    char              *unformatted_identity = NULL;
-    cJSON             *json_payload         = NULL;
-    char              *payload              = NULL;
-    char              *response             = NULL;
-    char              *signature            = NULL;
-    size_t             signature_length     = 0;
-    int                status               = 0;
+    char              *public_key_pem   = NULL;
+    mender_identity_t *identity         = NULL;
+    cJSON             *json_identity    = NULL;
+    cJSON             *json_payload     = NULL;
+    char              *payload          = NULL;
+    char              *response         = NULL;
+    char              *signature        = NULL;
+    size_t             signature_length = 0;
+    int                status           = 0;
 
     /* Get public key in PEM format */
     if (MENDER_OK != (ret = mender_tls_get_public_key_pem(&public_key_pem))) {
         mender_log_error("Unable to get public key");
         goto END;
     }
 
-    /* Get identity */
-    if (MENDER_OK != (ret = api_config.identity_cb(&identity))) {
-        mender_log_error("Unable to get identity");
-        goto END;
-    }
+    if (NULL == identity_info) {
```

Comment:
As we discussed in the past, the caching on the identity data is a behavior difference from the classic client. I understand the purpose, but this may hide problems during an update. Example:

An update module faulty implements something in the install state where the identity would change. This module does not require reboot. The update will continue and successfully commit. On next boot the device will be unauthenticated.

Reply:
> The update will continue and successfully commit.

This will happen even if we only cache authentication info, right? Because the JWT token will still be valid and the client will still be "authenticated".

Not caching authentication info would be quite foolish, of course. But should we drop the JWT token when an update doesn't require reboot so that it re-authenticates as part of its verification?

Reply:
> This will happen even if we only cache authentication info, right? Because the JWT token will still be valid and the client will still be "authenticated".

But this is also wrong on its own. Before commit we have to force a re-authentication for this very same reason: not to commit and then run into unauthenticated state. See [here](https://github.com/mendersoftware/mender/blob/4.0.4/src/mender-update/daemon/states.cpp#L669).

> But should we drop the JWT token when an update doesn't require reboot so that it re-authenticates as part of its verification?

Exactly. Although I would do it from the commit state code regardless of requiring a reboot or not.

Reply:
> Although I would do it from the commit state code regardless of requiring a reboot or not.

The commit state commits the update (i.e. does `mender_flash_confirm_image()` unconditionally), there's no check. The authentication check we have is in the `VERIFY_REBOOT` state which is only entered for updates that require reboot. For those that don't, `INSTALL` goes directly to `COMMIT` with no check.

Reply:
I added a function to trash the JWT. But I still don't understand where we should call it and where we should check that we can re-authenticate.

Reply:
I looked now at the code now,

> The commit state commits the update (i.e. does mender_flash_confirm_image() unconditionally), there's no check. 

Here with this reasoning you are talking about the implementation details of `zephyr-image` module, not our state machine.

What I am suggesting is to add `mender_api_drop_authentication_data()` and `mender_api_ensure_authenticated()` to the state machine code, inside `MENDER_UPDATE_STATE_COMMIT` state. If it fails here, it will go into rollback and eventually failure path.

What problems do you see doing it this way?

This is the closest to the design of the Mender C++ client, where the explicit re-auth is done in the state machine, not in the update module. This is important because we don't control the implementation of the update module.

> The authentication check we have is in the VERIFY_REBOOT state which is only entered for updates that require reboot. For those that don't, INSTALL goes directly to COMMIT with no check.

This is how our `zephyr-image` module is designed, but other modules may not care about being authenticated on `VERIFY_REBOOT` or have other requirements to check that the reboot was successful.

Reply:
> What I am suggesting is to add `mender_api_drop_authentication_data()` and `mender_api_ensure_authenticated()` to the state machine code, inside `MENDER_UPDATE_STATE_COMMIT` state. If it fails here, it will go into rollback and eventually failure path.
> 
> What problems do you see doing it this way?

Well, nothing apart from my generic feeling that authentication should not be the measure of an update being good or not. I can totally imagine cases where the server (or network) is not available all the time and rejecting an update just because there was, for example, a short network outage after rebooting into it doesn't sound right. It can be a totally valid approach, but isn't that a decision to be made by the update module developer? I'll add a commit implementing this, but my gut feeling is that it should be possible to disable at build time, at least.

Reply:
The explicit re-authentication in the classic Mender client is a painful lesson based on real worlds cases. Not having this catch can brick a device. To fail an update unnecessarily is preferable to commit a potentially broken update. And to determine that an update is valid we take the connectivity to the Mender server as a requirement.

Reply:
Okey, optional sounds good :memo: